Source/Core/Core/IOS/USB/Emulated/WiiSpeak.cpp

```diff
@@ -0,0 +1,412 @@
+// Copyright 2024 Dolphin Emulator Project
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#include "Core/IOS/USB/Emulated/WiiSpeak.h"
+
+#include <algorithm>
+
+#include "Core/Config/MainSettings.h"
+#include "Core/HW/Memmap.h"
+
+namespace IOS::HLE::USB
+{
+WiiSpeak::WiiSpeak(IOS::HLE::EmulationKernel& ios) : m_ios(ios)
+{
+  m_id = u64(m_vid) << 32 | u64(m_pid) << 16 | u64(9) << 8 | u64(1);
+}
+
+WiiSpeak::~WiiSpeak() = default;
+
+DeviceDescriptor WiiSpeak::GetDeviceDescriptor() const
+{
+  return m_device_descriptor;
+}
+
+std::vector<ConfigDescriptor> WiiSpeak::GetConfigurations() const
+{
+  return m_config_descriptor;
+}
+
+std::vector<InterfaceDescriptor> WiiSpeak::GetInterfaces(u8 config) const
+{
+  return m_interface_descriptor;
+}
+
+std::vector<EndpointDescriptor> WiiSpeak::GetEndpoints(u8 config, u8 interface, u8 alt) const
+{
+  return m_endpoint_descriptor;
+}
+
+bool WiiSpeak::Attach()
+{
+  if (m_device_attached)
+    return true;
+
+  DEBUG_LOG_FMT(IOS_USB, "[{:04x}:{:04x}] Opening device", m_vid, m_pid);
+  if (!m_microphone)
+    m_microphone = std::make_unique<Microphone>(m_sampler);
+  m_device_attached = true;
+  return true;
+}
+
+bool WiiSpeak::AttachAndChangeInterface(const u8 interface)
+{
+  if (!Attach())
+    return false;
+
+  if (interface != m_active_interface)
+    return ChangeInterface(interface) == 0;
+
+  return true;
+}
+
+int WiiSpeak::CancelTransfer(const u8 endpoint)
+{
+  INFO_LOG_FMT(IOS_USB, "[{:04x}:{:04x} {}] Cancelling transfers (endpoint {:#x})", m_vid, m_pid,
+               m_active_interface, endpoint);
+
+  return IPC_SUCCESS;
+}
+
+int WiiSpeak::ChangeInterface(const u8 interface)
+{
+  DEBUG_LOG_FMT(IOS_USB, "[{:04x}:{:04x} {}] Changing interface to {}", m_vid, m_pid,
+                m_active_interface, interface);
+  m_active_interface = interface;
+  return 0;
+}
+
+int WiiSpeak::GetNumberOfAltSettings(u8 interface)
+{
+  return 0;
+}
+
+int WiiSpeak::SetAltSetting(u8 alt_setting)
+{
+  return 0;
+}
+
+int WiiSpeak::SubmitTransfer(std::unique_ptr<CtrlMessage> cmd)
+{
+  DEBUG_LOG_FMT(IOS_USB,
+                "[{:04x}:{:04x} {}] Control: bRequestType={:02x} bRequest={:02x} wValue={:04x}"
+                " wIndex={:04x} wLength={:04x}",
+                m_vid, m_pid, m_active_interface, cmd->request_type, cmd->request, cmd->value,
+                cmd->index, cmd->length);
+
+  // Without a proper way to reconnect the emulated Wii Speak,
+  // this error after being raised prevents some games to use the microphone later.
+  //
+  // if (!IsMicrophoneConnected())
+  //  return IPC_ENOENT;
+
+  switch (cmd->request_type << 8 | cmd->request)
+  {
+  case USBHDR(DIR_DEVICE2HOST, TYPE_STANDARD, REC_INTERFACE, REQUEST_GET_INTERFACE):
+  {
+    const std::array<u8, 1> data{1};
```

Comment:
Personally I would use a simple `u8` here and then use the `&` operator in place of `.data()`. But I suppose an `std::array` is fine too if you prefer it.